sdk/typescript/src/transactions/ObjectCache.ts

```diff
@@ -0,0 +1,348 @@
+
+				if (cached.initialSharedVersion && !input.UnresolvedObject.initialSharedVersion) {
+					input.UnresolvedObject.initialSharedVersion = cached.initialSharedVersion;
+				}
+
+				if (cached.version && !input.UnresolvedObject.version) {
+					input.UnresolvedObject.version = cached.version;
+				}
+
+				if (cached.digest && !input.UnresolvedObject.digest) {
+					input.UnresolvedObject.digest = cached.digest;
+				}
+			}
+
+			await Promise.all(
+				blockData.transactions.map(async (tx) => {
+					if (tx.MoveCall) {
+						const def = await this.getMoveFunctionDefinition({
+							package: tx.MoveCall.package,
+							module: tx.MoveCall.module,
+							function: tx.MoveCall.function,
+						});
+
+						if (def) {
+							tx.MoveCall.argumentTypes = def.parameters;
+						}
+					}
+				}),
+			);
+
+			await next();
+
+			await Promise.all(
+				blockData.transactions.map(async (tx) => {
+					if (tx.MoveCall?.argumentTypes) {
+						await this.#cache.addMoveFunctionDefinition({
+							package: tx.MoveCall.package,
+							module: tx.MoveCall.module,
+							function: tx.MoveCall.function,
+							parameters: tx.MoveCall.argumentTypes,
+						});
+					}
+				}),
+			);
+		};
+	}
+
+	async clear() {
+		await this.#cache.clear();
+	}
+
+	async getMoveFunctionDefinition(ref: { package: string; module: string; function: string }) {
+		return this.#cache.getMoveFunctionDefinition(ref);
+	}
+
+	async getObjects(ids: string[]) {
+		return this.#cache.getObjects(ids);
+	}
+
+	async clearOwnedObjects() {
+		await this.#cache.clear('OwnedObject');
+	}
+
+	async applyEffects(effects: typeof bcs.TransactionEffects.$inferType) {
+		if (!effects.V2) {
+			throw new Error(`Unsupported transaction effects version ${effects.$kind}`);
+		}
+
+		const { lamportVersion, changedObjects } = effects.V2;
+
+		await Promise.all(
+			changedObjects.map(async ([id, change]) => {
+				if (change.outputState.NotExist) {
+					await this.#cache.deleteObject(id);
+				} else if (change.outputState.ObjectWrite) {
+					const [digest, owner] = change.outputState.ObjectWrite;
+
+					// Remove objects not owned by address after transaction
+					if (owner.ObjectOwner || (owner.AddressOwner && owner.AddressOwner !== this.#address)) {
+						await this.#cache.deleteObject(id);
+					}
+
+					await this.#cache.addObject({
+						objectId: id,
+						digest,
+						version: lamportVersion,
+						owner: owner.AddressOwner ?? owner.ObjectOwner ?? null,
+						initialSharedVersion: owner.Shared?.initialSharedVersion ?? null,
+					});
+				}
+			}),
+		);
+	}
+}
+
+export class CachingTransactionBlockExecutor {
+	#client: SuiClient;
+	cache: ObjectCache;
+
+	constructor({
+		client,
+		...options
+	}: ObjectCacheOptions & {
+		client: SuiClient;
+	}) {
+		this.#client = client;
+		this.cache = new ObjectCache(options);
+	}
+
+	/**
+	 * Clears all Owned objects
+	 * Immutable objects, Shared objects, and Move function definitions will be preserved
+	 */
+	async reset() {
```

Comment:
Not sure this is a good name.  The idea is to clear things that might have changed due to a transaction that happened outside the executor (just ownedObjects for now)